dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/NewSMSListener.java

```diff
@@ -0,0 +1,370 @@
+        sendSMSResponse( resp, sms, header.getSubmissionID() );
+    }
+
+    private void checkUser( SMSSubmission subm )
+    {
+        UID userid = subm.getUserID();
+        User user = userService.getUser( userid.uid );
+
+        if ( user == null )
+        {
+            throw new SMSProcessingException( SMSResponse.INVALID_USER.set( userid ) );
+        }
+    }
+
+    private SMSSubmissionHeader getHeader( IncomingSms sms )
+    {
+        byte[] smsBytes = SmsUtils.getBytes( sms );
+        SMSSubmissionReader reader = new SMSSubmissionReader();
+        try
+        {
+            return reader.readHeader( smsBytes );
+        }
+        catch ( Exception e )
+        {
+            e.printStackTrace();
+            Log.error( e.getMessage() );
+            return null;
+        }
+    }
+
+    private SMSMetadata getMetadata( Date lastSyncDate )
+    {
+        SMSMetadata meta = new SMSMetadata();
+        meta.dataElements = getAllDataElements( lastSyncDate );
+        meta.categoryOptionCombos = getAllCatOptionCombos( lastSyncDate );
+        meta.users = getAllUserIds( lastSyncDate );
+        meta.trackedEntityTypes = getAllTrackedEntityTypeIds( lastSyncDate );
+        meta.trackedEntityAttributes = getAllTrackedEntityAttributeIds( lastSyncDate );
+        meta.programs = getAllProgramIds( lastSyncDate );
+        meta.organisationUnits = getAllOrgUnitIds( lastSyncDate );
+
+        return meta;
+    }
+
+    private List<SMSMetadata.ID> getAllUserIds( Date lastSyncDate )
+    {
+        List<User> users = userService.getAllUsers();
+
+        return users.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllTrackedEntityTypeIds( Date lastSyncDate )
+    {
+        List<TrackedEntityType> teTypes = trackedEntityTypeService.getAllTrackedEntityType();
+
+        return teTypes.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllTrackedEntityAttributeIds( Date lastSyncDate )
+    {
+        List<TrackedEntityAttribute> teiAttributes = trackedEntityAttributeService.getAllTrackedEntityAttributes();
+
+        return teiAttributes.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllProgramIds( Date lastSyncDate )
+    {
+        List<Program> programs = programService.getAllPrograms();
+
+        return programs.stream().map( o -> getIdFromMetadata( o, lastSyncDate ) ).filter( Objects::nonNull )
+            .collect( Collectors.toList() );
+    }
+
+    private List<SMSMetadata.ID> getAllOrgUnitIds( Date lastSyncDate )
+    {
+        List<OrganisationUnit> orgUnits = organisationUnitService.getAllOrganisationUnits();
```

Comment:
Getting all OrganisationUnits is quite a heavy operation. We can move the logic to SQL level in order to fetch only those which are needed.